### PR TITLE
offer a basic initialization routine for test cases.

### LINKED ISSUE #9: Testground: Initialize sync client, net client and capture global + group seqs in `runner.Invoke`
## Context

Testground users have requested to move the most common scaffolding into the SDK itself. Since we'll be hammering out a lot of tests at Project Oni, it is a reasonable thing to do as it will make our codebase more DRY.

## Output

A PR in `testground/testground` that does what the title says.

### LINKED ISSUE #976: make `runtime.Invoke` perform a basic initialization ceremony
There is an argument to be made that all testplans have to wait for the data network to be initialised. We haven't document this very well currently and don't have great examples in the docs website. Maybe we should prebundle this with `runtime.Invoke`?

cc @aschmahmann

### PR DESCRIPTION
**Pull request by raulk**
This commit generifies `Invoke()` and `InvokeMap()` functions to take any of two test case function signatures:

- `TestCaseFn` (original)
- `InitializedTestCaseFn` (new)

When the latter is passed, we perform a basic initialization ceremony, which is recurringly seen across test plans. Refer to godocs of that type for more info.

This commit also moves the `runtime.Invoke*` methods into the `run` package. This was necessary to resolve an import cycle (thanks Go!).

Finally, this commit makes the functions in the `network` package no-op if we're running in a sidecar-less environment.

---

Closes https://github.com/testground/testground/issues/976.
Closes https://github.com/filecoin-project/oni/issues/9

### CODE REVIEW

File: run/invoker.go [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+	GlobalSeq  int64
+	GroupSeq   int64
+}
+
+// init can be safely invoked on a nil reference.
+func (ic *InitContext) init(runenv *runtime.RunEnv) {
+	var (
+		grpstate  = sync.State(fmt.Sprintf(StateInitializedGroupFmt, runenv.TestGroupID))
+		client    = sync.MustBoundClient(context.Background(), runenv)
+		netclient = network.NewClient(ic.SyncClient, runenv)
```
Comment (nonsense) [inline run/invoker.go]: Not sure how this is going to work on a `nil` `ic` - seems like we define `ic.SyncClient` below - maybe change to `client`, or?

File: run/invoker.go [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+
+type TestCaseFn func(env *runtime.RunEnv) error
+
+// InitializedTestCaseFn allows users to indicate they want a basic
+// initialization routine to be run before yielding control to the test case
+// function itself.
+//
+// The initialization routine is common scaffolding that gets repeated across
+// the test plans we've seen. We package it here in an attempt to keep your
+// code try.
```
Comment (nonsense) [inline run/invoker.go]: `dry`.

File: network/address.go [hunk 3]
```diff
@@ -13,7 +13,8 @@ import (
 func (c *Client) GetDataNetworkIP() (net.IP, error) {
 	re := c.runenv
 	if !re.TestSidecar {
-		return nil, ErrNoTrafficShaping
+		// this is a local runner; return the loopback address.
```
Comment (nonsense) [inline network/address.go]:
```suggestion
		// this must be a local:exec runner and we currently don't support traffic shaping on it
		// for now, just return the loopback address
```

I suggest a slightly more verbose comment as this is really a hack to do some kind of `data network` no-op.

File: network/client.go [hunk 4]
```diff
@@ -59,7 +59,9 @@ func (c *Client) MustWaitNetworkInitialized(ctx context.Context) {
 // either when the sidecar signals back to us, or when the context expires.
 func (c *Client) ConfigureNetwork(ctx context.Context, config *Config) (err error) {
 	if !c.runenv.TestSidecar {
-		return fmt.Errorf("failed to configure network; running in a sidecar-less environment")
+		msg := "ignoring network change request; running in a sidecar-less environment"
+		c.runenv.SLogger().Named("netclient").Warnf(msg)
```
Comment (nonsense) [inline network/client.go]:
```suggestion
		c.runenv.SLogger().Named("netclient").Warn(msg)
```

`msg` is not a template, so I am not sure if this won't complain (some loggers do).

File: run/invoker.go (hunk 2 above)
Comment (raulk) [reply run/invoker.go]: spell check... :D

File: network/address.go (hunk 3 above)
Comment (raulk) [reply network/address.go]: sure!

File: network/client.go (hunk 4 above)
Comment (raulk) [reply network/client.go]: done

File: run/invoker.go (hunk 1 above)
Comment (raulk) [reply run/invoker.go]: Oh yeah, good catch; I shuffled some code around, and this happened.

**Review by nonsense — APPROVED** [review]
LGTM

**Pull request merged**